### ISSUE TITLE
Load NULL in database as numpy.nan in pandas

### ISSUE THREAD

**Issue by zeromh** [capitalone/Data-Load-and-Copy-using-Python#69]
NULL values in Snowflake are loaded into python as `None`. When using `Database.to_dataframe`, these values remain as `None` rather than being converted to `numpy.nan`. As a result, any column containing `None` is forced to an "object" data type.

This makes it difficult to validate our data, since the data type has already changed. It also necessitates an extra step for type conversion.

I haven't tested this fully, but this issue could be fixed by changing [this line](https://github.com/capitalone/Data-Load-and-Copy-using-Python/blob/master/locopy/database.py#L204)
```fetched = [tuple(column for column in row) for row in fetched]```

 to something like:
```
fetched = [tuple(column if column is not None else np.nan for column in row) 
    for row in fetched]
```

`np.nan` seems to be the de facto null value to use with pandas, as it doesn't mess up one's dtypes. If `Database.to_dataframe` is meant to be a convenient way of porting data into pandas, I think it makes sense for the method to be aware of the issue with `None` and to handle nulls more gracefully.

**Comment by fdosani** [comment]
I think this makes sense. Let me take a closer look once I’m back to work
in early January. Happy to incorporate this sort of functionality into the
to pandas function.

**Comment by jborchma** [comment]
I think this makes sense, but might lead to a different not-wanted behaviour. If there is an integer column with missing values, it will be cast as float (new behaviour) instead of object (current behaviour). This could lead to issues when there are integer values used as unique identifiers, for example. There is [this](https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html) functionality now in pandas, but it needs to be invoked manually and won't be inferred. Not sure if this is something we should leverage as well (it would mean pandas >= 0.24.0, we support currently >=0.19.0)

**Comment by zeromh** [comment]
I'm not sure I'm following you. Are you describing a case where someone depends on an int column to load into pandas as an object, because they believe that the int column will have some nulls? (Without nulls an int column would load as int.)

That seems like a bad practice to support. Not sure if I'm misunderstanding you though.

**Comment by jborchma** [comment]
No, not necessarily. I just wanted to bring up that with this change we might introduce some new, unexpected behaviour.

**Comment by fdosani** [comment]
I'm looking to make a new release and wondering if this is something we can roll into that? Maybe we can iterate on this next week and figure out the scope of work here, and if it make sense to include it?

**Comment by fdosani** [comment]
Circling back on this. So to @jborchma point and link he referenced, it is a valid concern where ints might be cast unintentionally into floats if we convert `NULL` --> `nan`. But the upside is having `nan` for actual missing values which is probably better for numeric computation. I need to think about this a bit more, as this could be a breaking change for the users here. Thoughts?

**Comment by fdosani** [comment]
Was messing around with some scenarios:

Example where `id` is an int. it seems to force to a float in this case.
```
[(123457, 'Amanda', 'Panda', None),
 (123456, 'Joe', 'Blow', datetime.date(2008, 1, 1)),
 (123457, 'Amanda', 'Panda', None),
 (123457, 'Amanda', 'Panda', None),
 (None, 'Amanda', 'Panda', None),
 (123459, None, 'Amanda', None),
 (1234599, 'Amanda', None, None)]

In [19]: z = pandas.DataFrame(fetched, columns=columns)                                                                                                                                                                                                               

In [20]: z                                                                                                                                                                                                                                                            
Out[20]: 
     acct_id first_nm last_nm date_of_birth
0   123457.0   Amanda   Panda          None
1   123456.0      Joe    Blow    2008-01-01
2   123457.0   Amanda   Panda          None
3   123457.0   Amanda   Panda          None
4        NaN   Amanda   Panda          None
5   123459.0     None  Amanda          None
6  1234599.0   Amanda    None          None

In [21]: z.dtypes                                                                                                                                                                                                                                                     
Out[21]: 
acct_id          float64
first_nm          object
last_nm           object
date_of_birth     object
dtype: object
```

If the `id` is a decimal it will coerce into a `object`:
```
[(Decimal('123456.5'), 'Joe', 'Blow', datetime.date(2008, 1, 1)),
 (Decimal('123457.0'), 'Amanda', 'Panda', None),
 (None, 'Amanda', 'Panda', None),
 (Decimal('123459.0'), None, 'Amanda', None),
 (Decimal('1234599.0'), 'Amanda', None, None)]

In [28]: z = pandas.DataFrame(fetched, columns=columns)                                                                                                                                                                                                               

In [29]: z                                                                                                                                                                                                                                                            
Out[29]: 
     acct_id first_nm last_nm date_of_birth
0   123456.5      Joe    Blow    2008-01-01
1   123457.0   Amanda   Panda          None
2       None   Amanda   Panda          None
3   123459.0     None  Amanda          None
4  1234599.0   Amanda    None          None

In [30]: z.dtypes                                                                                                                                                                                                                                                     
Out[30]: 
acct_id          object
first_nm         object
last_nm          object
date_of_birth    object
dtype: object
```

TLDR; I'm starting to think we maybe shouldn't be touching this. This is just a convenience function to get something quickly into pandas. The coercing is a best guess and not something we should be trying to figure out. I'm leaning towards users transforming their own data, as the intent was more to remove the boilerplate stuff with loading and unloading data from SF and RS.

**Comment by zeromh** [comment]
I confirmed with some Snowflake data that "number" columns with nulls end up in pandas as `float`, rather than `object`. So there must be something funky with my OP table such that my numeric column turned into 'object'.

I think there might not actually be an issue here.

**Comment by fdosani** [comment]
I will close out this issue for now. Maybe we revisit it down the road or if others encounter a problem or a pressing need arises.
